content/existential-containers.md

```diff
@@ -0,0 +1,288 @@
+
+def largest(xs: Seq[AnyPolygon]): Option[AnyPolygon] =
+  xs.maxByOption((a) => a.witness.area(a.value))
+```
+
+The type `AnyPolygon` conceptually represents an arbitrary polygon.
+It consists of a pair containing some arbitrary value as well as a witness of that value's type being a polygon.
+We call this pair an _existential container_, as a nod to a similar feature in Swift.
+
+While the above example "hardcodes" the type class `Polygon`, Scala's type system is actually rich enough to safely define a generic abstraction represeting existential containers parameterized by a type class.
+A possible implementation is presented in the appendix but its details are actually irrelevant to for the proposed change.
+The purpose of this SIP is _only_ to support the selection of an existential container's `value` field implicitly.
+That way, one could simply write `xs.maxByOption(_.area)` in the above example, resulting in quite idiomatic scala.
+
+To illustrate further, assume the existence of an abstraction named `Containing[TC]` for representing containers pairing an arbitrary value with a witness of its conformance to some type class `TC`.
+The proposed change would let the compiler accept the following example:
+
+```scala
+trait Polygon extends TypeClass:
+  extension (self: Self) def area: Double
+
+def largest(xs: Seq[Containing[Polygon]]): Option[Containing[Polygon]] =
+  xs.maxByOption(_.area)
+```
+
+This implementation of `largest` requires existential containers to take and return arbitrary polygons.
+Indeed, we wish to operate on a _heterogeneous_ list of polygons (i.e., types conforming to `Polygon`), not a list of a particular type happing to have an instance of the type class.
+On the return side, we wish to return any type known to be a polygon paired with the witness of its conformance.
+Again, doing so (conveniently) is not possible without existential containers.
+
+The above example generalizes to any occurrence of heterogeneous collection.
+For instance:
+
+```scala
+trait CustomHashable extends TypeClass:
+  extension (self: Self) def hashInto(hasher: Hasher)
+
+def customHashValue(xs: List[Containing[CustomHashable]]): Int =
+  val h = Hasher()
+  for x <- xs do xs.hashInto(h)
+  h.finalize()
+```
+
+Returning a value paired with its witness generalizes similarly.
+
+```scala
+trait Sizeable extends TypeClass:
+  extension (self: Self) def size: Int
+
+def shortest[A: Sizeable, B: Sizeable](a: A, b: B): Containing[Sizeable] =
+  if b.size < a.size then Containing(a) else Containing(b)
+```
+
+Further motivation for existential containers in Scala have been described in a research paper [2].
+
+### Specification
+
+Assuming the existence of an abstraction named `Containing[TC]` for representing containers pairing an arbitrary value with a witness of its conformance to some type class `TC` in the standard library, the compiler injects the selection of the `value` field implicitly when a method of `Containing[TC]` is selected.
```

Comment:
Since this is in the `Specification`section, can I assume that "the compiler injects ..." is in fact part of what this SIP specifies? If so, it should be fleshed out:
- How does this interact with other adaptations, including implicit conversions and apply insertion?
- In which condition does this adaptation triggers exactly? Is Containing treated specially by the compiler or can we (and should we) generalize this?

Otherwise, this should be in a separate section (example or appendix)